hack/Makefile.debug

```diff
@@ -2,29 +2,137 @@
 
-push:
+push-image:
 	$(IMAGEBUILDER) tag $(IMAGE):$(TAG) $(REGISTRY)/$(IMAGE):$(TAG)
 	$(IMAGEBUILDER) push $(REGISTRY)/$(IMAGE):$(TAG)
 
 set-image:
 	oc scale --replicas 0 -n openshift-cluster-version deployments/cluster-version-operator
 	oc scale --replicas 0 -n openshift-ingress-operator deployments ingress-operator
-	oc -n openshift-ingress scale deployment --replicas=0 router-default
-	oc -n openshift-ingress patch deployment router-default -p '{"spec":{"template":{"spec":{"$$setElementOrder/containers":[{"name":"router"}],"containers":[{"imagePullPolicy":"Always","name":"router"}]}}}}'
+	oc scale --replicas 0 -n openshift-ingress deployment/router-default
+ifeq ($(DOCKERFILE_DEBUG),hack/remote-debug/Dockerfile)
+	oc apply -f hack/remote-debug/manifests
+	oc apply -f hack/remote-debug/custom-router-scc.yaml
+	@for patch in $(sort $(wildcard hack/remote-debug/patches/deployment/router-default/*.yaml)); do \
+	    echo "Applying patch $$patch"; \
+	    oc -n openshift-ingress patch deployment router-default -p "$$(cat $$patch)" --type=strategic; \
+	done
+endif
 	oc -n openshift-ingress set image deployment/router-default router=$(REGISTRY)/$(IMAGE):$(TAG)
-	oc -n openshift-ingress scale deployment --replicas=1 router-default
+	oc scale --replicas 1 -n openshift-ingress deployment/router-default
+ifeq ($(DOCKERFILE_DEBUG),hack/remote-debug/Dockerfile)
+	@echo "Waiting for the deployment/router-default to become available..."; \
+	if ! oc -n openshift-ingress rollout status deployment/router-default --timeout=180s; then \
+		echo "Deployment did not become available in time. Please check the deployment status."; \
+		false; \
+	fi
+	@max_attempts=10; \
+	attempt=1; \
+	while [ $$attempt -le $$max_attempts ]; do \
+		echo "Checking SSH access..."; \
+		if ssh -o ConnectTimeout=5 -F hack/remote-debug/ssh_config container uptime; then \
+			echo "SSH check passed."; \
+			break; \
+		else \
+			echo "SSH check failed. Restarting port-forward services and retrying..."; \
+			$(MAKE) -f hack/Makefile.debug restart-port-forward-services; \
+			attempt=$$((attempt + 1)); \
+			sleep 5; \
+		fi; \
+	done; \
+	if [ $$attempt -gt $$max_attempts ]; then \
+		echo "SSH check did not pass in time. Please check the deployment and SSH configuration."; \
+		false; \
+	fi
+endif
 
-dwim:  new-openshift-router-image push set-image
+deploy-image: build-image push-image set-image
 
 reset:
 	oc scale --replicas 1 -n openshift-cluster-version deployments/cluster-version-operator
 	oc scale --replicas 1 -n openshift-ingress-operator deployments ingress-operator
+	oc delete -n openshift-ingress deployment/router-default
+	oc delete --ignore-not-found -f hack/remote-debug/manifests
+	oc delete --ignore-not-found -f hack/remote-debug/custom-router-scc.yaml
+
+rsync-openshift-router: build
+	rsync --progress --compress -e 'ssh -F hack/remote-debug/ssh_config' ./openshift-router container:/usr/bin/openshift-router
+
+debug: rsync-openshift-router
```

Comment:
If you're happy for it to run each and every time.

Reply:
> Should we add `install-port-forward-services` as a pre-requisite for this target? Otherwise this command has to be run manually, right?

No, I don't think we should add this as a pre-requisite. We have the ssh check which should ensure that ssh connectivity is working. I don't want to add any delays between "I just build" and now "copy the binary to the container".

Reply:
> We have the ssh check which should ensure that ssh connectivity is working.

Interesting, in README, [`set-image` is suggested to be run](https://github.com/openshift/router/blob/ba8f66c29cc80b1cd59d93ca9f12d2d0f0b8439e/hack/remote-debug/README.md#building-and-deploying-the-debug-image)  before [`install-port-forward-services`](https://github.com/openshift/router/blob/ba8f66c29cc80b1cd59d93ca9f12d2d0f0b8439e/hack/remote-debug/README.md#preparatory-steps-1). Should we set `install-port-forward-services` as a prereq of `set-image` then?

Reply:
The steps seem disconnected. The port-forwarding service will always fail until the debug image is set, as there’s nothing listening on port 2222 before that. I approached the steps as: prepare the new image, set the new image, and then expect the port forwarding to work after install-port-forward-services has run.

Reply:
Done in 1b47bf547e8b05b9533c9dca1720dbb9e4a14075.